crates/wasi-common/src/table.rs

```diff
@@ -10,83 +12,89 @@ use std::collections::HashMap;
 
     /// Check if the table has a resource at the given index.
     pub fn contains_key(&self, key: u32) -> bool {
-        self.map.contains_key(&key)
+        self.map.read().unwrap().contains_key(&key)
     }
 
     /// Check if the resource at a given index can be downcast to a given type.
     /// Note: this will always fail if the resource is already borrowed.
     pub fn is<T: Any + Sized>(&self, key: u32) -> bool {
-        if let Some(r) = self.map.get(&key) {
+        if let Some(r) = self.map.read().unwrap().get(&key) {
             r.is::<T>()
         } else {
             false
         }
     }
 
-    /// Get an immutable reference to a resource of a given type at a given index. Multiple
-    /// immutable references can be borrowed at any given time. Borrow failure
-    /// results in a trapping error.
-    pub fn get<T: Any + Sized>(&self, key: u32) -> Result<&T, Error> {
-        if let Some(r) = self.map.get(&key) {
-            r.downcast_ref::<T>()
-                .ok_or_else(|| Error::badf().context("element is a different type"))
+    /// Get an Arc reference to a resource of a given type at a given index. Multiple
+    /// immutable references can be borrowed at any given time.
+    pub fn get<T: Any + Send + Sync + Sized>(&self, key: u32) -> Result<Arc<T>, Error> {
+        if let Some(r) = self.map.read().unwrap().get(&key).cloned() {
+            r.downcast::<T>()
+                .map_err(|_| Error::badf().context("element is a different type"))
         } else {
             Err(Error::badf().context("key not in table"))
         }
     }
 
-    /// Get a mutable reference to a resource of a given type at a given index. Only one mutable
-    /// reference can be borrowed at any given time. Borrow failure results in a trapping error.
-    pub fn get_mut<T: Any + Sized>(&mut self, key: u32) -> Result<&mut T, Error> {
-        if let Some(r) = self.map.get_mut(&key) {
-            r.downcast_mut::<T>()
-                .ok_or_else(|| Error::badf().context("element is a different type"))
-        } else {
-            Err(Error::badf().context("key not in table"))
-        }
+    /// Remove a resource at a given index from the table. Returns the resource
+    /// if it was present.
+    pub fn delete<T: Any + Send + Sync>(&self, key: u32) -> Option<Arc<T>> {
+        self.map
+            .write()
+            .unwrap()
+            .remove(&key)
+            .map(|r| r.downcast::<T>().unwrap())
     }
 
     /// Remove a resource at a given index from the table. Returns the resource
     /// if it was present.
-    pub fn delete(&mut self, key: u32) -> Option<Box<dyn Any + Send + Sync>> {
-        self.map.remove(&key)
+    pub fn renumber(&self, from: u32, to: u32) -> Result<(), Error> {
+        let from_entry = self
+            .map
+            .write()
+            .unwrap()
+            .remove(&from)
+            .ok_or(Error::badf())?;
+        self.map.write().unwrap().insert(to, from_entry);
+        Ok(())
```

Comment:
This should probably be an atomic operation rather than two seaparate ones.

Reply:
You are right! Thanks!